testplan/testing/environment.py

```diff
@@ -0,0 +1,269 @@
+                )
+        for d in self._resources.values():
+            if d.async_start is not UNSET:
+                raise ValueError(
+                    f"`async_start` parameter of driver {d} should not "
+                    "be set if driver dependency is specified."
+                )
+
+        self._dependency = dependency
+
+        # leverage PuLP as scheduler as well as cycle detector
+        self._problem = LpProblem("aha", LpMinimize)
+        self._variables: DriverVariableMap = {}
+
+        for d_uid in self._resources.keys():
+            self._variables[d_uid] = DriverVariable(d_uid)
+
+        # specify optimize objective
+        self._problem += lpSum([v.start for v in self._variables.values()])
+        for side_a, side_b in self._dependency.edges:
+            # add linear constraints
+            self._problem += self._gen_dep_constraint(side_a, side_b)
+
+        # use built-in lp solver
+        ret = self._problem.solve(PULP_CBC_CMD(msg=False, timeLimit=10))
+        # -1 for no solution exists, 0 for no solution found, 1 for solution found
+        if ret == -1:
+            raise ValueError(
+                "Unable to work out any schedule, possibly due to cyclic dependency."
+            )
+        if ret != 1:
+            raise RuntimeError(
+                "Internal driver scheduler error, "
+                f"solver reporting status {LpStatus[self._problem.status]}."
+            )
+
+    def _gen_dep_constraint(
+        self, side_a: "Driver", side_b: "Driver"
+    ) -> LpConstraint:
+        """We first play the Side-A of a cassette."""
+        return (
+            self._variables[side_a.uid()].start
+            + self._variables[side_a.uid()].elapsed
+            <= self._variables[side_b.uid()].start
+        )
+
+    def start_in_pool(self, *_):
+        raise RuntimeError(
+            "TestEnvironment.start_in_pool: Would not be invoked by design."
+        )
+
+    def stop_in_pool(self, *_):
+        raise RuntimeError(
+            "TestEnvironment.stop_in_pool: Would not be invoked by design."
+        )
+
+    def start(self):
+        """
+        Start the drivers either in the legacy way or following the driver dependency.
+        """
+        if self._dependency is None:
+            # we got no dependency declared, go with the legacy way,
+            # override `async_start` of drivers
+            for d in self._resources.values():
+                if d.async_start is UNSET:
+                    d.async_start = False
+            return super().start()
+
+        buckets = defaultdict(list)
+        for v in self._variables.values():
+            buckets[int(v.start.value())].append(v.driver_uid)
+
+        for k in sorted(buckets.keys()):
+            # we tend to have a consistant behaviour here
+            self._naive_batch_start_drivers(
+                self._resources[d_uid] for d_uid in sorted(buckets[k])
+            )
```

Comment:
I think this env start up is a dynamic process, that a driver shall be kick start once all its dependencies are started.